storage/innobase/log/log0recv.cc

```diff
@@ -1781,14 +1787,32 @@ static byte *recv_parse_or_apply_log_rec_body(
           /** else the index is flushed to disk before
           backup started hence no error */
         } else {
+          // While scaning redo logs during a backup operation a
+          // MLOG_INDEX_LOAD type redo log record indicates, that a DDL
+          // (create index, alter table...) is performed with
+          // 'algorithm=inplace'. If using pagetracking, the affected tablespace
+          // must be copied using full scan. Record it in the index_load_list
+          // We do this during the first scan of redo before starting file copy
+          // thread. This copy is required because Page-tracking relies on
+          // changes from redo as well. ie PXB has to copy redo as well apart
+          // from copying pages from the pages given by page-tracking list.
+          // Since in-place DDL skips redo logging and the pages that are not
+          // yet flushed, will not be tracked by page-tracking. And since the
+          // redo logging is skipped, PXB will fail to get those changed pages
+          // (currently only in buffer pool). hence, we rely on re-copying the
```

Comment:
> (currently only in buffer pool)

Is this true? if they are not redo logged and only reside on BP, what happens in case of a crash?
I believe they are already flushed to the tablespace, but we don't know which ones, otherwise what is the point of copying the .ibd if the changes are only on BP?

cc: @satya-bodapati

Reply:
I think I gave this comment to Rahul. So there is more to this story.
For all the changes skipped to redo log, they are only in BP. Only after complete flush to disk, we see the redo log record. So copying the IBD on seeing this redo is correct.

So the problematic case is this.  ALTER ADD INDEX  happens after the current checkpoint_lsn (backup has already started copying from this checkpoint_lsn). and the checkpoint_lsn doesn't move

In this case, all data is flushed to disk, redo is written, but the page-tracking data is outdated. We see the changes done by ALTER ADD INDEX from page-tracking, only on next checkpoint.

The main culprit is..well, no one.. actually a design choice. page-tracking data is flushed at checkpoints.

To look at another way, we don't need to copy IBDs for all the MLOG_INDEX_LOAD seen before the current checkpoint_lsn as seen by backup. Because those changed-pages are available at the start of backup (page-tracking flushed those at the checkpoint_lsn seen by us)

The problematic DDLs are only those MLOG_INDEX_LOADs after the checkpoint_lsn seen by backup and when checkpoint_lsn doesn't move.

It took me a while to understand this whole thing but I think I got it right :wink: Hope it is clear to you as well, @altmannmarcelo

Reply:
@satya-bodapati understood.

@rahulmalik87 can we adjust the `(parenthesis)` comment to reflect that they are indeed flushed to .idb file.